args_test.go

```diff
@@ -5,231 +5,208 @@ import (
+	o, e := executeCommand(newCommand(MinimumNArgs(2), true), "one", "three")
+	expectSuccess(o, e, t)
+}
+
+func TestMinimumNArgsWithValidWithInvalidArgs(t *testing.T) {
+	_, e := executeCommand(newCommand(MinimumNArgs(2), true), "a", "b")
+	expectError(e, t, "valid")
 }
 
 func TestMinimumNArgsWithLessArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: MinimumNArgs(2), Run: emptyRun}
-	_, err := executeCommand(c, "a")
+	_, e := executeCommand(newCommand(MinimumNArgs(2), false), "a")
+	expectError(e, t, "min")
+}
 
-	if err == nil {
-		t.Fatal("Expected an error")
-	}
+func TestMinimumNArgsWithLessArgsWithValid(t *testing.T) {
+	_, e := executeCommand(newCommand(MinimumNArgs(2), true), "one")
+	expectError(e, t, "min")
+}
 
-	got := err.Error()
-	expected := "requires at least 2 arg(s), only received 1"
-	if got != expected {
-		t.Fatalf("Expected %q, got %q", expected, got)
-	}
+func TestMinimumNArgsWithLessArgsWithValidWithInvalidArgs(t *testing.T) {
+	_, e := executeCommand(newCommand(MinimumNArgs(2), true), "a")
+	expectError(e, t, "valid")
 }
 
+// MaximumNArgs
+
 func TestMaximumNArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: MaximumNArgs(3), Run: emptyRun}
-	output, err := executeCommand(c, "a", "b")
-	if output != "" {
-		t.Errorf("Unexpected output: %v", output)
-	}
-	if err != nil {
-		t.Errorf("Unexpected error: %v", err)
-	}
+	o, e := executeCommand(newCommand(MaximumNArgs(3), false), "a", "b")
+	expectSuccess(o, e, t)
 }
 
-func TestMaximumNArgsWithMoreArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: MaximumNArgs(2), Run: emptyRun}
-	_, err := executeCommand(c, "a", "b", "c")
+func TestMaximumNArgsWithValid(t *testing.T) {
+	o, e := executeCommand(newCommand(MaximumNArgs(2), true), "one", "three")
+	expectSuccess(o, e, t)
+}
 
-	if err == nil {
-		t.Fatal("Expected an error")
-	}
+func TestMaximumNArgsWithValidWithInvalidArgs(t *testing.T) {
+	_, e := executeCommand(newCommand(MaximumNArgs(2), true), "a", "b")
+	expectError(e, t, "valid")
+}
 
-	got := err.Error()
-	expected := "accepts at most 2 arg(s), received 3"
-	if got != expected {
-		t.Fatalf("Expected %q, got %q", expected, got)
-	}
+func TestMaximumNArgsWithMoreArgs(t *testing.T) {
+	_, e := executeCommand(newCommand(MaximumNArgs(2), false), "a", "b", "c")
+	expectError(e, t, "max")
 }
 
-func TestExactArgs(t *testing.T) {
-	c := &Command{Use: "c", Args: ExactArgs(3), Run: emptyRun}
-	output, err := executeCommand(c, "a", "b", "c")
-	if output != "" {
-		t.Errorf("Unexpected output: %v", output)
-	}
-	if err != nil {
-		t.Errorf("Unexpected error: %v", err)
-	}
+func TestMaximumNArgsWithMoreArgsWithValid(t *testing.T) {
+	_, e := executeCommand(newCommand(MaximumNArgs(2), true), "one", "three", "two")
+	expectError(e, t, "max")
 }
 
-func TestExactArgsWithInvalidCount(t *testing.T) {
-	c := &Command{Use: "c", Args: ExactArgs(2), Run: emptyRun}
-	_, err := executeCommand(c, "a", "b", "c")
+func TestMaximumNArgsWithMoreArgsWithValidWithInvalidArgs(t *testing.T) {
```

Comment:
I understand you're trying to be very descriptive with these names but IMHO it's very difficult to read.

Reply:
This is fixed in #842, where underscores are introduced. E.g. `TestMinimumNArgsWithValidWithInvalidArgs` is renamed to `TestArgs_MinimumN_wValid_InvalidArgs`, which is explained as:

- `TestArgs`: we are testing the args
- `MinimumN`: this is the validator we are testing
- `wValid`: the command does contain a `ValidArgs` field. This can be `wArgs`, `wValid` or `wArgswValid`.
- `InvalidArgs`: execution arguments.

See: https://github.com/umarcor/cobra/compare/feat-matchvalid...umarcor:feat-validinhelp#diff-748c41438b4135ae7035c2af1228e349

I think that the format with underscores is both easy to read and useful. See output: https://travis-ci.com/umarcor/cobra/jobs/186549357

Note that using underscores in test function names does not produce any linting issue.

If you want, I can add the underscores in this PR too.

Reply:
The underscores do help but still long and hard to read. I think the tests could be greatly improved by implementing them as sub-tests and grouping them per validation function.  The nitty gritty descriptive parts then go in the subtests.

Perhaps an improvement for another PR.  For now, imma call the name fine since after looking back at the tests_args.go file it seems a lot of these names are too long including ones you haven't changes/added.

Reply:
> I think the tests could be greatly improved by implementing them as sub-tests and grouping them per validation function. The nitty gritty descriptive parts then go in the subtests.
> Perhaps an improvement for another PR. 

That sounds good. This is my first experience with golang tests, tho. So, I'd like to do those changes in a different PR (probably #842).

> For now, if you could add the underscores that would be great. After looking back at the args_test.go file it seems a lot of these names are too long including ones you haven't changes/added.

Done.

Reply:
In the end, I used sub-tests. I think that the implementation can be improved, because the items in the map are executed in random order and it would be desirable to group them according to the first 'field'. But I hope you find it better than the previous approach.